File: def_hierarchical_pathfinding/hierarchical_pathfinding.lua
-- Hierarchical Pathfinding Library for Defold
-- Main module that orchestrates chunk-based pathfinding

local coord_utils = require "src.utils.coord_utils"
local chunk_navigator = require "src.chunk_navigator"
local path_segment_builder = require "src.path_segment_builder"
local local_pathfinder = require "src.local_pathfinder"

local M = {}

-- Module state
local config = nil
local chunk_graph = nil
local path_cache = nil

-- Initialize the pathfinding system
-- @param init_config Configuration table with required fields:
--   chunk_size: Size of chunk in tiles
--   tile_size: Size of tile in world units
--   map_width: Map width in chunks
--   map_height: Map height in chunks
--   get_chunk_data: Function(chunk_id) -> chunk_data
--   transition_points: Array of transition point data
function M.init(init_config)
    -- Validate required fields
    assert(init_config, "Configuration required")
    assert(type(init_config.chunk_size) == "number" and init_config.chunk_size > 0, 
        "Invalid chunk_size")
    assert(type(init_config.tile_size) == "number" and init_config.tile_size > 0, 
        "Invalid tile_size")
    assert(type(init_config.map_width) == "number" and init_config.map_width > 0, 
        "Invalid map_width")
    assert(type(init_config.map_height) == "number" and init_config.map_height > 0, 
        "Invalid map_height")
    assert(type(init_config.get_chunk_data) == "function", 
        "get_chunk_data must be a function")
    assert(type(init_config.transition_points) == "table", 
        "transition_points must be a table")
    
    -- Store configuration
    config = init_config
    
    -- Generate list of all chunk IDs
    local all_chunks = {}
    for y = 0, config.map_height - 1 do
        for x = 0, config.map_width - 1 do
            table.insert(all_chunks, coord_utils.coords_to_chunk_id(x, y))
        end
    end
    
    -- Build chunk navigation graph
    chunk_graph = chunk_navigator.build_chunk_graph(config.transition_points, all_chunks)
    
    -- Initialize cache if enabled
    if config.enable_cache then
        path_cache = {}
        config.cache_size = config.cache_size or 100
    end
end

-- Find path from start to end position
-- @param start_pos Global start position {x, y, z}
-- @param end_pos Global end position {x, y, z}
-- @return Array of segments {chunk, position} or nil if no path
function M.find_path(start_pos, end_pos)
    assert(config, "Pathfinder not initialized. Call init() first.")
    assert(start_pos and start_pos.x and start_pos.y, "Invalid start position")
    assert(end_pos and end_pos.x and end_pos.y, "Invalid end position")
    
    -- Check cache
    if path_cache then
        local cache_key = string.format("%d,%d_%d,%d", 
            start_pos.x, start_pos.y, end_pos.x, end_pos.y)
        local cached = path_cache[cache_key]
        if cached then
            return cached.segments
        end
    end
    
    -- Get chunk IDs
    local start_chunk = coord_utils.global_to_chunk_id(start_pos, config.chunk_size, config.tile_size)
    local end_chunk = coord_utils.global_to_chunk_id(end_pos, config.chunk_size, config.tile_size)
    
    -- Check if positions are walkable
    local start_local = coord_utils.global_to_local(start_pos, start_chunk, config.chunk_size, config.tile_size)
    local end_local = coord_utils.global_to_local(end_pos, end_chunk, config.chunk_size, config.tile_size)
    
    local start_chunk_data = config.get_chunk_data(start_chunk)
    local end_chunk_data = config.get_chunk_data(end_chunk)
    
    if not start_chunk_data or not end_chunk_data then
        return nil
    end
    
    if not local_pathfinder.is_walkable(start_chunk_data, start_local) or
       not local_pathfinder.is_walkable(end_chunk_data, end_local) then
        return nil
    end
    
    -- Find chunk path
    local chunk_path = chunk_navigator.find_chunk_path(chunk_graph, start_chunk, end_chunk)
    if not chunk_path then
        return nil
    end
    
    -- Build path segments
    local segments = path_segment_builder.build_segments(
        chunk_path, start_pos, end_pos, config
    )
    
    -- Cache result
    if path_cache and segments then
        local cache_key = string.format("%d,%d_%d,%d", 
            start_pos.x, start_pos.y, end_pos.x, end_pos.y)
        
        -- Simple LRU: remove oldest if cache full
        local cache_count = 0
        for _ in pairs(path_cache) do
            cache_count = cache_count + 1
        end
        
        if cache_count >= config.cache_size then
            -- Remove first found (simple eviction)
            for k in pairs(path_cache) do
                path_cache[k] = nil
                break
            end
        end
        
        path_cache[cache_key] = {
            segments = segments,
            timestamp = os.time()
        }
    end
    
    return segments
end

-- Helper function to get chunk ID from global position
-- @param global_pos Global position
-- @return Chunk ID string
function M.get_chunk_from_global(global_pos)
    assert(config, "Pathfinder not initialized")
    return coord_utils.global_to_chunk_id(global_pos, config.chunk_size, config.tile_size)
end

-- Convert global position to local within chunk
-- @param global_pos Global position
-- @param chunk_id Chunk ID
-- @return Local position {x, y}
function M.global_to_local(global_pos, chunk_id)
    assert(config, "Pathfinder not initialized")
    return coord_utils.global_to_local(global_pos, chunk_id, config.chunk_size, config.tile_size)
end

-- Convert local position to global
-- @param local_pos Local position within chunk
-- @param chunk_id Chunk ID
-- @return Global position {x, y, z}
function M.local_to_global(local_pos, chunk_id)
    assert(config, "Pathfinder not initialized")
    return coord_utils.local_to_global(local_pos, chunk_id, config.chunk_size, config.tile_size)
end

-- Check if position is walkable
-- @param global_pos Global position
-- @return boolean
function M.is_position_walkable(global_pos)
    assert(config, "Pathfinder not initialized")
    
    local chunk_id = M.get_chunk_from_global(global_pos)
    local chunk_data = config.get_chunk_data(chunk_id)
    
    if not chunk_data then
        return false
    end
    
    local local_pos = M.global_to_local(global_pos, chunk_id)
    return local_pathfinder.is_walkable(chunk_data, local_pos)
end

-- Check if two positions can reach each other
-- @param start_pos Start position
-- @param end_pos End position
-- @return boolean
function M.can_reach(start_pos, end_pos)
    local segments = M.find_path(start_pos, end_pos)
    return segments ~= nil
end

-- Clear path cache
function M.clear_cache()
    if path_cache then
        path_cache = {}
    end
end

return M 

File: library_example/hierarchical_pathfinding/HierarchicalPathfinding.js
/**
 * Hierarchical Pathfinding Library for JavaScript
 * Minimal version with pre-computed connections graph
 */

import { CoordUtils } from './src/utils/CoordUtils.js';
import { TransitionGraph } from './src/TransitionGraph.js';
import { LocalPathfinder } from './src/LocalPathfinder.js';

export class HierarchicalPathfinding {
    constructor() {
        this.transitionGraph = null;
        this.config = null;
    }

    /**
     * Initialize the pathfinding system
     * @param {Object} config - Configuration containing:
     *   - tileSize: tile size (in world units)
     *   - gridWidth/gridHeight: grid dimensions (in chunks)
     *   - chunkWidth/chunkHeight: chunk dimensions (in tiles)
     *   - getChunkData: function returning chunk data
     *   - transitionPoints: array of transition points between chunks
     */
    init(config) {
        // Basic validation
        if (!config || !config.tileSize || 
            !config.gridWidth || !config.gridHeight || 
            !config.chunkWidth || !config.chunkHeight ||
            !config.getChunkData || !config.transitionPoints) {
            throw new Error("Missing required configuration parameters");
        }
        
        // Save configuration
        this.config = config;
        
        // Build connections graph between transition points
        this.transitionGraph = new TransitionGraph(config.transitionPoints, {
            gridWidth: config.gridWidth,
            gridHeight: config.gridHeight,
            chunkSize: config.chunkWidth, // Use chunkWidth as chunkSize for compatibility
            tileSize: config.tileSize
        });
    }

    /**
     * Main function - finds path from startPos to endPos
     * @param {Object} startPos - Start position {x, y} in world units
     * @param {Object} endPos - End position {x, y} in world units
     * @returns {Array|null} - Array of segments [{chunk, position}] or null
     */
    findPath(startPos, endPos) {
        if (!this.config) {
            throw new Error("Pathfinder has not been initialized");
        }

        // Check if positions are within world bounds
        const worldWidth = this.config.gridWidth * this.config.chunkWidth * this.config.tileSize;
        const worldHeight = this.config.gridHeight * this.config.chunkHeight * this.config.tileSize;
        
        if (startPos.x < 0 || startPos.x >= worldWidth || 
            startPos.y < 0 || startPos.y >= worldHeight ||
            endPos.x < 0 || endPos.x >= worldWidth ||
            endPos.y < 0 || endPos.y >= worldHeight) {
            return null;
        }

        // Determine which chunks contain start and end
        const startChunk = CoordUtils.globalToChunkId(startPos, this.config.chunkWidth, this.config.tileSize);
        const endChunk = CoordUtils.globalToChunkId(endPos, this.config.chunkWidth, this.config.tileSize);

        // If same chunk - try local A* first
        if (startChunk === endChunk) {
            const localPath = this.findLocalPath(startChunk, startPos, endPos);
            // If local path was found, return it immediately
            if (localPath) {
                return localPath;
            }
            // If not, allow continuation to hierarchical search.
            // It may happen that points are in the same chunk, but in separate,
            // unconnected areas, so we need to go outside.
        }

        // Different chunks (or same chunk without local path) - search through transition points
        const startPoint = this.findNearestTransition(startPos, startChunk);
        const endPoint = this.findNearestTransition(endPos, endChunk);

        if (!startPoint || !endPoint) {
            return null; // No available transition points
        }
        
        const transitionPath = this.transitionGraph.findPath(startPoint.id, endPoint.id);

        if (!transitionPath) {
            return null; // No path between chunks
        }

        // Build final path segments
        const segments = this.buildPathSegments(startPos, endPos, transitionPath);
        
        return segments;
    }

    /**
     * Finds nearest transition point in a given chunk
     * @param {Object} pos - Position for which we're looking for a point
     * @param {string} chunkId - Chunk ID
     * @returns {Object|null} - Nearest available transition point
     */
    findNearestTransition(pos, chunkId) {
        // Get all transition points in this chunk
        const points = this.transitionGraph.getPointsInChunk(chunkId);
        
        if (points.length === 0) {
            return null;
        }

        let nearest = null;
        let minDistance = Infinity;

        // Find nearest point that can be reached
        for (const point of points) {
            // Calculate global position of transition point
            const pointPos = CoordUtils.getTransitionGlobalPosition(
                point, chunkId, this.config.chunkWidth, this.config.tileSize
            );

            if (!pointPos) {
                continue;
            }

            // Check if we can reach this point with local path
            const localPath = this.findLocalPath(chunkId, pos, pointPos);

            if (localPath) {
                // Calculate Euclidean distance
                const dx = pointPos.x - pos.x;
                const dy = pointPos.y - pos.y;
                const distance = Math.sqrt(dx * dx + dy * dy);
                
                if (distance < minDistance) {
                    minDistance = distance;
                    nearest = point;
                }
            }
        }

        return nearest;
    }

    /**
     * Finds local path within a single chunk
     * @param {string} chunkId - Chunk ID
     * @param {Object} startPos - Start position (global)
     * @param {Object} endPos - End position (global)
     * @returns {Array|null} - Path segment or null
     */
    findLocalPath(chunkId, startPos, endPos) {
        // Get chunk data (2D array)
        const chunkData = this.config.getChunkData(chunkId);
        if (!chunkData) {
            return null;
        }

        // Convert global positions to local positions in chunk
        const localStart = CoordUtils.globalToLocal(startPos, chunkId, this.config.chunkWidth, this.config.tileSize);
        const localEnd = CoordUtils.globalToLocal(endPos, chunkId, this.config.chunkWidth, this.config.tileSize);

        // Find path with local A*
        const localPath = LocalPathfinder.findPath(chunkData, localStart, localEnd);

        if (localPath) {
            // Return as single segment
            return [{
                chunk: chunkId,
                position: endPos
            }];
        }

        return null;
    }

    /**
     * Finds chunk for connection between two transition points
     * @param {Object} fromPoint - Start point
     * @param {Object} toPoint - Target point  
     * @returns {string|null} - Chunk ID or null if no connection
     */
    findConnectionChunk(fromPoint, toPoint) {
        const connection = fromPoint.connections.find(conn => conn.id === toPoint.id);
        return connection ? connection.chunk : null;
    }

    /**
     * Returns clean transition points with redundant node optimization
     * @param {Object} startPos - Start position 
     * @param {Object} endPos - End position  
     * @param {Array} transitionPath - List of transition point IDs
     * @returns {Array} - List of transition points
     */
    buildPathSegments(startPos, endPos, transitionPath) {
        const segments = [];
        
        // Determine start and end chunks
        const startChunk = CoordUtils.globalToChunkId(startPos, this.config.chunkWidth, this.config.tileSize);
        const endChunk = CoordUtils.globalToChunkId(endPos, this.config.chunkWidth, this.config.tileSize);
        
        // If no transition points (direct path)
        if (transitionPath.length === 0) {
            segments.push({
                chunk: startChunk,
                position: endPos
            });
            return segments;
        }
        
        // Create copy of path for optimization
        let effectivePath = [...transitionPath];
        
        // 🔥 FIRST NODE VERIFICATION
        if (effectivePath.length >= 2) {
            const firstPoint = this.transitionGraph.getPoint(effectivePath[0]);
            const secondPoint = this.transitionGraph.getPoint(effectivePath[1]);
            
            // Check if second point is accessible from start chunk
            // AND if there's a direct connection from first to second point in this chunk.
            const connectionChunk = this.findConnectionChunk(firstPoint, secondPoint);
            if (secondPoint.chunks.includes(startChunk) && connectionChunk === startChunk) {
                effectivePath.shift(); // Remove first, redundant node
            }
        }
        
        // Add start segment (from startPos to first transition point)
        if (effectivePath.length > 0) {
            const firstPoint = this.transitionGraph.getPoint(effectivePath[0]);
            const firstPointPos = CoordUtils.getTransitionGlobalPosition(
                firstPoint, startChunk, this.config.chunkWidth, this.config.tileSize
            );
            
            

            if (firstPointPos) {
                segments.push({
                    chunk: startChunk,
                    position: firstPointPos
                });
            }
        }
        
        // Build segments between transition points
        for (let i = 0; i < effectivePath.length - 1; i++) {
            const currentPoint = this.transitionGraph.getPoint(effectivePath[i]);
            const nextPoint = this.transitionGraph.getPoint(effectivePath[i + 1]);
            
            // Use helper function to find connection chunk
            const connectionChunk = this.findConnectionChunk(currentPoint, nextPoint);
            
            if (!connectionChunk) {
                continue;
            }
            
            const nextPointPos = CoordUtils.getTransitionGlobalPosition(
                nextPoint, connectionChunk, this.config.chunkWidth, this.config.tileSize
            );

            if (!nextPointPos) {
                continue;
            }
            
            segments.push({
                chunk: connectionChunk,
                position: nextPointPos
            });
        }
        
        // Add end segment (from last transition point to endPos)
        if (effectivePath.length > 0) {
            segments.push({
                chunk: endChunk,
                position: endPos
            });
        }

        // 🔥 PENULTIMATE SEGMENT VERIFICATION
        if (segments.length >= 2) {
            const penultimateSegment = segments[segments.length - 2];
            
            if (penultimateSegment.chunk === endChunk) {
                segments.splice(segments.length - 2, 1); // Remove penultimate segment
            }
        }
        
        return segments;
    }
} 